ugs-platform/ugs-platform-plugin-pendant/src/main/java/com/willwinder/ugs/platform/plugin/pendant/PendantServer.java

```diff
@@ -0,0 +1,212 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with UGS.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.willwinder.ugs.platform.plugin.pendant;
+
+import com.google.gson.Gson;
+import com.willwinder.ugs.nbp.lib.lookup.CentralLookup;
+import com.willwinder.universalgcodesender.listeners.ControllerStateListener;
+import com.willwinder.universalgcodesender.listeners.ControllerStatus;
+import com.willwinder.universalgcodesender.listeners.UGSEventListener;
+import com.willwinder.universalgcodesender.model.BackendAPI;
+import com.willwinder.universalgcodesender.model.UGSEvent;
+import java.io.ByteArrayOutputStream;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import net.glxn.qrgen.QRCode;
+import net.glxn.qrgen.image.ImageType;
+import org.webbitserver.BaseWebSocketHandler;
+import org.webbitserver.WebServer;
+import org.webbitserver.WebServers;
+import org.webbitserver.WebSocketConnection;
+
+public class PendantServer extends BaseWebSocketHandler implements UGSEventListener, ControllerStateListener {
+
+    private static final Logger logger = Logger.getLogger(PendantServer.class.getName());
+
+    private final BackendAPI backend;
+    private int port = 9988;
+
+    private WebServer server;
+    private final List<WebSocketConnection> connections = new ArrayList<>();
+    private final PendantCommandHandler commandHandler;
+
+    public PendantServer() {
+        backend = CentralLookup.getDefault().lookup(BackendAPI.class);
+        backend.addUGSEventListener(this);
+        backend.addControllerStateListener(this);
+        commandHandler = new PendantCommandHandler();
+    }
+
+    /**
+     * When a new web socket connection from a client is established.
+     *
+     * @param connection
+     */
+    @Override
+    public void onOpen(WebSocketConnection connection) {
+        connections.add(connection);
+        connection.send(buildStateString(null));
+    }
+
+    /**
+     * When a client terminates a connection.
+     *
+     * @param connection
+     */
+    @Override
+    public void onClose(WebSocketConnection connection) {
+        logger.info("Pendant client disconnected");
+    }
+
+    /**
+     * When a client sends a web socket message.
+     *
+     * @param connection
+     * @param message
+     */
+    @Override
+    public void onMessage(WebSocketConnection connection, String message) {
+        commandHandler.handleCommand(message);
+    }
+
+    /**
+     * Called by the GUI Component to start the server.
+     *
+     * @throws Exception
+     */
+    public void start() throws Exception {
+        server = WebServers.createWebServer(port);
+
+//        server.add("/", new StaticFileHandler("./"));
```

Comment:
It looks like webbit might not support serving files in the classpath. I'll have to investigate what utilities exist for extracting files into the netbeans cache directories. I also manually extract resource files in other places but that's more of a plan B.

If you could include the static assets in a resource directory I'll do some experimenting.

Reply:
I was trying to use the EmbeddedResourceHandler first from
https://github.com/webbit/webbit/blob/master/src/main/java/org/webbitserver/handler/EmbeddedResourceHandler.java

but beyond that I couldn't even get an http response back from it.

Reply:
Oh perfect. Java can be a PITA with how you format your classpath string, I can help investigate that.